### PR TITLE
FEATURE: Add new NodePrivilege `isAncestorNodeOf`

### PR DESCRIPTION
**Pull request by johannessteu**
This adds a new NodePrivilege `isAncestorNodeOf` which allows to define easily a whitelist approach for the node tree

### CODE REVIEW

**Comment by johannessteu** [comment]
Hey @aertmann, 
this is what @bwaidelich and me came up with on a quick try. The code right now would be ok to create a read only approach for ancestor nodes, but not for editing nodes because the sublings are not respected yet

**Comment by johannessteu** [comment]
This is a policy.yaml to test with:

```
privilegeTargets:

  'TYPO3\Neos\Security\Authorization\Privilege\NodeTreePrivilege':
    NodeTree.Whitelist:
      matcher: 'TRUE'

    NodeTree.TEST:
      matcher: 'isAncestorNodeOf("/sites/neosdemotypo3org/features/navigation-elements/first-item")'

    NodeTree.TEST2:
      matcher: 'isDescendantNodeOf("/sites/neosdemotypo3org/features/navigation-elements/first-item")'

  'TYPO3\TYPO3CR\Security\Authorization\Privilege\Node\EditNodePrivilege':
    NodeTree.TEST3:
      matcher: 'isAncestorNodeOf("/sites/neosdemotypo3org/features/navigation-elements/first-item")'

roles:

  'TYPO3.Neos:Administrator':
    privileges:
      -
        privilegeTarget: 'NodeTree.TEST'
        permission: GRANT
      -
        privilegeTarget: 'NodeTree.TEST2'
        permission: GRANT
      -
        privilegeTarget: 'NodeTree.TEST3'
        permission: DENY
```

What do we want to achieve in general:
- A easy configuration to allow a editor to work in a subtree, and only in this subtree
- all ancestorNoes should be read only
- the given node and its children should be editable

The Policy forces a whitelost approach. First of we restrict all nodes to abstain for `NodeTree` and `EditNode`. Then we allow all ancestor nodes of `first-item` for the tree and all sub nodes as well. This does work right now and is huge improvement compared to the current state.
The  `TYPO3\TYPO3CR\Security\Authorization\Privilege\Node\EditNodePrivilege` then restrict all ancestor nodes for editing.

**Comment by kdambekalns** [comment]
Hint: Travis builds can be restarted from the build page, either as whole or for individual jobs. :)

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Security/Authorization/Privilege/Node/NodePrivilegeContext.php [hunk 1]
```diff
@@ -64,6 +64,31 @@ public function setNode(NodeInterface $node)
     }
 
     /**
+     * @param $nodePathOrIdentifier
```
Comment (daniellienert) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Security/Authorization/Privilege/Node/NodePrivilegeContext.php]: The type is missing

**Comment by bwaidelich** [comment]
05ddf43 tweaks the `NodePrivilegeContext` a little bit. Behat tests run through for me.
I'm still confused by the behavior of the whole node related ACL..: For the very common scenario of giving an editor edit access to a specific node that is *not* on the 1st level is still cumbersome..
I'd expect the following to work:

```
privilegeTargets:

  'TYPO3\Neos\Security\Authorization\Privilege\NodeTreePrivilege':
    'My.Site:NodeTree.All':
      matcher: 'TRUE'
    'My.Site:NodeTree.Team1':
      matcher: 'isAncestorOrDescendantNodeOf("/sites/my-site/teams/team1")'

  'TYPO3\TYPO3CR\Security\Authorization\Privilege\Node\EditNodePrivilege':
    'My.Site:Edit.All':
      matcher: 'TRUE'
    'My.Site:Edit.Team1':
      matcher: 'isDescendantNodeOf("/sites/my-site/teams/team1")'
```

to work if I grant an editor access to `My.Site:NodeTree.Team1` and `My.Site:Edit.Team1`: He should be able to navigate to the `team1` page but disallowed to edit parents. But edit controls on parent pages are not disabled for that role and you only get an exception when trying to publish because you'll edit nodes that are not on the path..
IMO the issue is that `NodeTreePrivilege` extends `EditNodePrivilege` - if that's not the case it seems easier to control the ACL

**Comment by bwaidelich** [comment]
Addition: IMO we should adjust https://neos.readthedocs.io/en/stable/CreatingASite/Security.html with an example setup with this change

**Comment by hhoechtl** [comment]
I just tested it with different scenarios and it works great.
The most simple configuration for this is:
```yaml
privilegeTargets:
  'TYPO3\Neos\Security\Authorization\Privilege\NodeTreePrivilege':
    NodeTree.Whitelist:
      matcher: 'TRUE'

    NodeTree.TEST:
      matcher: 'isAncestorOrDescendantNodeOf("/sites/neosdemo/features/navigation-elements")'

  'TYPO3\TYPO3CR\Security\Authorization\Privilege\Node\EditNodePrivilege':
    NodeTree.WhiteList:
      matcher: 'TRUE'
    NodeTree.TEST3:
      matcher: 'isAncestorNodeOf("/sites/neosdemo/features/navigation-elements")'

roles:
  'TYPO3.Neos:Administrator':
    privileges:
      -
        privilegeTarget: 'NodeTree.TEST'
        permission: GRANT
      -
        privilegeTarget: 'NodeTree.TEST3'
        permission: DENY
```

It works like charm and even the remark of @bwaidelich that the edit controls of parent pages are not disabled seems to have been fixed by some other commit.

**Comment by johannessteu** [comment]
Ready for review & merge from my side if the tests are green!

**Comment by bwaidelich** [comment]
very nice, thanks for reviewing!
IMO https://github.com/neos/neos-development-collection/pull/477#issuecomment-238529896 still applies, maybe I can take care of that tomorrow

**Comment by johannessteu** [comment]
Yeah, sorry. Didn't mentioned that. We wanted to create a new section with example setups for different use cases. But we should tackle this in a separate PR as this is not only related to this change

**Comment by bwaidelich** [comment]
Alright then, thanks for the input

**Pull request merged**